src/duckling-tests/state/undo-redo.spec.ts

```diff
@@ -0,0 +1,227 @@
+            this.store.dispatch(testAction({update : 1}));
+            this.store.dispatch(redoAction());
+            expect(getState(this.store).update).toBe(1);
+        });
+
+        it("redoes a single undo", function() {
+            this.store.dispatch(testAction({update : 1}));
+            this.store.dispatch(undoAction());
+            this.store.dispatch(redoAction());
+            expect(getState(this.store).update).toBe(1);
+        });
+
+        it("can redo multiple undos", function() {
+            this.store.dispatch(testAction({update : 1}));
+            this.store.dispatch(testAction({update : 2}));
+            this.store.dispatch(testAction({update : 3}));
+
+            this.store.dispatch(undoAction());
+            this.store.dispatch(undoAction());
+            this.store.dispatch(undoAction());
+
+            this.store.dispatch(redoAction());
+            this.store.dispatch(redoAction());
+
+            expect(getState(this.store).update).toBe(2);
+        });
+    });
+
+    describe("mergeKey", function() {
+        it("unique merge keys do not merge", function() {
+            this.store.dispatch(testAction({update : 1}, 1));
+            this.store.dispatch(testAction({update : 2}, 2));
+
+            this.store.dispatch(undoAction());
+
+            expect(getState(this.store).update).toBe(1);
+        });
+
+        it("identical merge keys do merge", function() {
+            this.store.dispatch(testAction({update : 1}, 1));
+            this.store.dispatch(testAction({update : 2}, 2));
+            this.store.dispatch(testAction({update : 3}, 2));
+            this.store.dispatch(testAction({update : 4}, 3));
+
+            this.store.dispatch(undoAction());
+            this.store.dispatch(undoAction());
+
+            expect(getState(this.store).update).toBe(1);
+        });
+
+        it("merges redo operations", function() {
+            this.store.dispatch(testAction({update : 1}, 1));
+            this.store.dispatch(testAction({update : 2}, 2));
+            this.store.dispatch(testAction({update : 3}, 2));
+            this.store.dispatch(testAction({update : 4}, 3));
+
+            this.store.dispatch(undoAction());
+            this.store.dispatch(undoAction());
+
+            this.store.dispatch(redoAction());
+
+            expect(getState(this.store).update).toBe(3);
+        });
+
+        it("is ignored after a redo", function() {
+            this.store.dispatch(testAction({update : 1}, 1));
+            this.store.dispatch(testAction({update : 2}, 2));
+            this.store.dispatch(testAction({update : 3}, 2));
+
+            this.store.dispatch(undoAction());
+            this.store.dispatch(redoAction());
+
+            this.store.dispatch(testAction({update : 4}, 2));
+
+            this.store.dispatch(undoAction());
+
+            expect(getState(this.store).update).toBe(3);
+        });
+    });
+
+    describe("autoMerger", function() {
```

Comment:
Should this be nested in the undo-redo test since it's altering the store just like that one does?